### PR TITLE
Only use GlobalPermissionRegistry

### LINKED ISSUE #99: Better enforce erc20 limits
Due to contracts like safe erc20 we cant reliably enforce our limits in the wallet scheme.
Not sure if this is something we would consider fixing as we don't really consider these additional implementations of erc20 throughout the code since they are not really the standard. 
But perhaps we should discuss this. 

Screenshot 2022-03-28 at 09 55 19

### PR DESCRIPTION
**Pull request by AugustoL**
fix #99 #100 #101 

- Remove PermissionRegistry smart contract in favor of using GlobalPermissionRegistry for all dxdao-contracts.
- Make GlobalPermissionRegistry ownable and allow owner to set any permission.
- Upgrade solidity compiler version in WalletScheme to 0.8.8 without changing solidity compiler version for daostack smart contracts. This is why I removed the imports and I communicate with the daostack contracts and voting machine using encoded  function calls.
- Adapt tests to work with the changes done in the contracts, minimal changes had to be done and I even fixed some tests that were not checking the conditions they were supposed to check.

The good thing about using GlobalPermissionRegistry is that the checks of the permissions are done in the permission registry itself by telling the permission registry the value sent, it also allowed me to simplify a lot the WalletScheme code and remove some variables and permissions check that were a mess. https://github.com/DXgovernance/dxdao-contracts/pull/111/files#diff-e968da9277857e3ed6886fee3104db6d19821825eb98743ecd57f5a8c3c8d50aL158

### CODE REVIEW

File: contracts/dxvote/WalletScheme.sol [hunk 1]
```diff
@@ -466,4 +400,132 @@ contract WalletScheme is DXDVotingMachineCallbacks, ProposalExecuteInterface {
...
+    }
+
+    mapping(bytes32 => uint256) public proposalsBlockNumber;
+
+    function mintReputation(
+        uint256 _amount,
+        address _beneficiary,
+        bytes32 _proposalId
+    ) external onlyVotingMachine returns (bool) {
+        // return ControllerInterface(avatar.owner()).mintReputation(_amount, _beneficiary, address(avatar));
```
Comment (Kenny-Gin1) [inline contracts/dxvote/WalletScheme.sol:420]: 🧹

File: contracts/dxvote/WalletScheme.sol [hunk 2]
```diff
@@ -466,4 +400,132 @@ contract WalletScheme is DXDVotingMachineCallbacks, ProposalExecuteInterface {
...
+            );
+    }
+
+    function stakingTokenTransfer(
+        IERC20 _stakingToken,
+        address _beneficiary,
+        uint256 _amount,
+        bytes32 _proposalId
+    ) external onlyVotingMachine returns (bool) {
+        // return ControllerInterface(avatar.owner()).externalTokenTransfer(_stakingToken, _beneficiary, _amount, avatar);
```
Comment (Kenny-Gin1) [inline contracts/dxvote/WalletScheme.sol:463]: 🧹

File: contracts/dxvote/WalletScheme.sol [hunk 3]
```diff
@@ -466,4 +400,132 @@ contract WalletScheme is DXDVotingMachineCallbacks, ProposalExecuteInterface {
...
+                        address(avatar)
+                    ),
+                    "WalletScheme: DXDVotingMachine callback externalTokenTransfer error"
+                ),
+                (bool)
+            );
+    }
+
+    function getNativeReputation() public view returns (address) {
+        // return Avatar(avatar).nativeReputation();
```
Comment (Kenny-Gin1) [inline contracts/dxvote/WalletScheme.sol:481]: 🧹

File: contracts/dxvote/WalletScheme.sol [hunk 4]
```diff
@@ -466,4 +400,132 @@ contract WalletScheme is DXDVotingMachineCallbacks, ProposalExecuteInterface {
...
+                (uint256)
+            );
+    }
+
+    function balanceOfStakingToken(IERC20 _stakingToken, bytes32 _proposalId) external view returns (uint256) {
+        return _stakingToken.balanceOf(address(avatar));
+    }
+
+    function getTotalReputationSupply(bytes32 _proposalId) external view returns (uint256) {
+        // return Avatar(avatar).nativeReputation().totalSupplyAt(proposalsBlockNumber[_proposalId]);
```
Comment (Kenny-Gin1) [inline contracts/dxvote/WalletScheme.sol:509]: 🧹

File: contracts/dxvote/WalletScheme.sol [hunk 5]
```diff
@@ -466,4 +400,132 @@ contract WalletScheme is DXDVotingMachineCallbacks, ProposalExecuteInterface {
...
+                avatar.functionStaticCall(
+                    abi.encodeWithSignature("nativeReputation()"),
+                    "WalletScheme: DXDVotingMachine callback nativeReputation error"
+                ),
+                (address)
+            );
+    }
+
+    function getNativeReputationTotalSupply() public view returns (uint256) {
+        // return Avatar(avatar).nativeReputation().totalSupply();
```
Comment (Kenny-Gin1) [inline contracts/dxvote/WalletScheme.sol:493]: 🧹

File: test/dxvote/PermissionRegistry.js [hunk 6]
```diff
@@ -221,22 +180,25 @@ contract("PermissionRegistry", function (accounts) {
...
     );
-    const testCallAllowedFromTime = (
-      await permissionRegistry.getPermission(
-        constants.NULL_ADDRESS,
-        quickWalletScheme.address,
-        actionMock.address,
-        callData.substring(0, 10)
-      )
-    ).fromTime;
+    // const testCallAllowedFromTime = (
```
Comment (Kenny-Gin1) [inline test/dxvote/PermissionRegistry.js:183]: 🧹

File: test/dxvote/PermissionRegistry.js [hunk 7]
```diff
@@ -252,15 +214,15 @@ contract("PermissionRegistry", function (accounts) {
     const proposalId2 = await helpers.getValueFromLogs(tx2, "_proposalId");
 
     // The call to execute is not allowed YET, because we change the delay time to 30 seconds
-    await expectRevert(
-      votingMachine.contract.vote(proposalId2, 1, 0, constants.NULL_ADDRESS, {
-        from: accounts[2],
-      }),
-      "call not allowed"
-    );
+    // await expectRevert(
```
Comment (Kenny-Gin1) [inline test/dxvote/PermissionRegistry.js:217]: 🧹

File: contracts/dxvote/WalletScheme.sol [hunk 8]
```diff
@@ -359,10 +280,23 @@ contract WalletScheme is DXDVotingMachineCallbacks, ProposalExecuteInterface {
...
+            controller.functionStaticCall(
+                abi.encodeWithSignature("getSchemeParameters(address,address)", address(this), avatar),
+                "WalletScheme: getSchemeParameters error"
+            ),
+            (bytes32)
+        );
 
         // Get the proposal id that will be used from the voting machine
-        bytes32 proposalId = votingMachine.propose(2, voteParams, msg.sender, address(avatar));
+        // bytes32 proposalId = votingMachine.propose(2, voteParams, msg.sender, address(avatar));
```
Comment (Kenny-Gin1) [inline contracts/dxvote/WalletScheme.sol:292]: 🧹

**Review by Kenny-Gin1 — COMMENTED** [review]
Nice job! Looks cleaner now indeed. Where do we set the address for `votingMachine`?

**Comment by AugustoL** [comment]
@Kenny-Gin1 I left those comments in the smart contracts as a reference to the smart contract call that was executing with the imported daostack contracts, later in the next versions when we upgrade daostack contracts solidty versions we have to use again the commented code.

And in the tests I commented some parts that we have to change to find a way to test that conditions again and dont loose coverage.

I think we can leave these comments for now.

**Review by rossneilson — APPROVED** [review]

**Pull request merged**